Add Total Bunch Attributes to Monitor

ax3l opened the pull request:
Add our reduced beam characteristics (moments of the particle bunch) as metadata to openPMD files (the beam monitor element).

X-Ref.:
- #533

## Action Items

- [x] implement
- [x] test that `reduced_beam_characteristics` is indeed an AllReduce now.
- [x] document changes in reduced beam characteristics (no ref particle info anymore)
- [x] document openPMD attributes
- [x] test and use the new metadata from openPMD over the text file in tests (only `visualize_ml_....py`?) @ax3l @RTSandberg 

## Follow-ups

- [ ] document openPMD particle "columns" if read back to a dataframe @ax3l #618

ax3l committed:
Add Total Bunch Attributes to Monitor

Add our reduced beam diagnostics (moments of the particle bunch) as
metadata to openPMD files (the beam monitor element).

n01r reviewed (approved):
Looks good. That's really useful! :)

ax3l committed:
Update Reduce Beam Diagnostics

Remove ref particle quantities.
Update examples and docs.
Add beta and gamma consistently to ref particle output.

ax3l commented:
Will push one more update on the TODOs we marked in the RBC routines - removing reference particle info from it to avoid duplication and collisions.

ax3l commented:
ping @RTSandberg @n01r for review :)

cemitch99 reviewed (commented):
Just to clarify regarding the reference particle:  
The reference particle data is still written to "ref_particle_," but is no longer duplicated in "reduced_beam_characteristics_".  Instead, it is duplicated in the openPMD output.  

Is this the intent?  Did I miss anything?

RTSandberg commented on examples/pytorch_surrogate_model/visualize_ml_surrogate_15_stage.py:
```diff
@@ -291,14 +291,15 @@ def plot_beam_df(
     "diags/reduced_beam_characteristics.*"
 )
-ref_gamma = np.sqrt(1 + impactx_surrogate_reduced_diags["ref_beta_gamma"] ** 2)
+ref_gamma = np.sqrt(1 + impactx_surrogate_ref_particle["ref_beta_gamma"] ** 2)
```
```suggestion
ref_gamma = impactx_surrogate_ref_particle["gamma"]
```

RTSandberg commented on examples/pytorch_surrogate_model/visualize_ml_surrogate_15_stage.py:
```diff
@@ -291,14 +291,15 @@ def plot_beam_df(
     ref_gamma
     - impactx_surrogate_reduced_diags["pt_mean"]
-    * impactx_surrogate_reduced_diags["ref_beta_gamma"]
+    * impactx_surrogate_ref_particle["ref_beta_gamma"]
```
```suggestion
    * impactx_surrogate_ref_particle["beta_gamma"]
```

RTSandberg reviewed (changes requested):
This looks good to me; I added a couple revisions to the 15 stages example and it runs for me

ax3l committed:
Finalize: Vis Surrogate Updates

Co-authored-by: Ryan Sandberg <RSandberg@lbl.gov>

ax3l commented:
> The reference particle data is still written to "ref_particle_," but is no longer duplicated in "reduced_beam_characteristics_".

Correct. The issue was that it would otherwise collide in names once we combine it to the openPMD output.
If we want, we can still keep the information in the text file output.

> Instead, it is duplicated in the openPMD output.

It should now only be once in the openPMD output using a consistent name for reference particle and collective beam properties; stored as attributes to the beam in each step (iteration).

ax3l commented on src/particles/elements/diagnostics/openPMD.cpp:
```diff
@@ -265,8 +267,14 @@ namespace detail
+
+        // total particle bunch information
+        //   @see impactx::diagnostics::reduced_beam_characteristics
+        for(const auto &kv : m_rbc) {
```
```suggestion
        for (const auto &kv : m_rbc) {
```

ax3l committed:
Formatting

RTSandberg reviewed (approved):
Looks good to me!

cemitch99 reviewed (commented):
I think it's important that we keep the reference particle text output.  It is a small cost for I/O, and allows a quick check by the user without any postprocessing.

n01r reviewed (approved):
Looks good to me :)

ax3l commented:
> I think it's important that we keep the reference particle text output. It is a small cost for I/O, and allows a quick check by the user without any postprocessing.

Good point. We have currently in `diags/` the
- `ref_particle*`
- `reduced_beam_characteristics*`

In the latter, this would remove `s` and `ref_beta_gamma` now. Shall I add `s` again (or both?).
(Later, we could also make both one file for simplicity.)

ax3l committed:
Text Output of RBC: Keep `s` column

Position on the beamline/ring is important for quick processing.

cemitch99 commented:
I think it's good to remove `ref_beta_gamma` from `reduced_beam_characteristics`, since that information is duplicated in the `ref_particle` text output, and it is a "reference" quantity.  However, it would be good to keep the `s` output in both text files, since this is effectively the independent variable characterizing the lattice location.

ax3l commented:
Perfect, ok that is what I thought, too. I added the `s` column back to the text file now :)
